internal/service/resources/collector/utils.go

```diff
@@ -0,0 +1,225 @@
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to update object '%s/%s': %w", object.TableName(), uuid, err)
+		}
+
+		slog.Debug("object updated",
+			"table", object.TableName(), "uuid", uuid, "before", before, "after", after, "columns", tags.Fields())
+	default:
+		return nil, nil, fmt.Errorf("failed to get object '%s/%s': %w", object.TableName(), uuid, err)
+	}
+
+	return before, after, nil
+}
+
+// persistDataChangeEvent persists a data change object to its database table.  The before and
+// after model objects are marshaled to JSON prior to being stored.
+func persistDataChangeEvent(ctx context.Context, db *pgxpool.Pool, tableName string, uuid uuid.UUID,
+	parentUUID *uuid.UUID, before, after any) (*models.DataChangeEvent, error) {
+	var err error
+	var beforeJSON, afterJSON []byte
+	if before != nil {
+		if beforeJSON, err = json.Marshal(before); err != nil {
+			return nil, fmt.Errorf("failed to marshal before object: %w", err)
+		}
+	}
+	if after != nil {
+		if afterJSON, err = json.Marshal(after); err != nil {
+			return nil, fmt.Errorf("failed to marshal after object: %w", err)
+		}
+	}
+
+	dataChangeEvent := models.DataChangeEvent{
+		ObjectType: tableName,
+		ObjectID:   uuid,
+		ParentID:   parentUUID,
+	}
+
+	if beforeJSON != nil {
+		value := string(beforeJSON)
+		dataChangeEvent.BeforeState = &value
+	}
+	if afterJSON != nil {
+		value := string(afterJSON)
+		dataChangeEvent.AfterState = &value
+	}
+
+	result, err := utils.Create[models.DataChangeEvent](ctx, db, dataChangeEvent)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create data change event: %w", err)
+	}
+
+	return result, nil
+}
+
+// persistObjectWithChangeEvent persists an object to its database table and if the external API
+// model representation of the object has changed then a data change event is stored.  Persisting
+// of the object and its change event are captured under the same transaction to ensure we never
+// lose any change events.
+func persistObjectWithChangeEvent[T db.Model](ctx context.Context, db *pgxpool.Pool, record T,
+	uuid uuid.UUID, parentUUID *uuid.UUID,
+	converter GenericModelConverter) (*models.DataChangeEvent, error) {
+	var dataChangeEvent *models.DataChangeEvent
+	txCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	tx, err := db.Begin(txCtx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to begin transaction: %w", err)
+	}
+
+	before, after, err := persistObject(ctx, db, record, uuid)
+	if err != nil {
+		return nil, fmt.Errorf("failed to persist object: %w", err)
+	}
+
+	afterModel := converter(*after)
+	var beforeModel any = nil
+	if before != nil {
+		value := converter(*before)
+		beforeModel = value
+	}
+
+	if beforeModel == nil || !reflect.DeepEqual(beforeModel, afterModel) {
+		// Capture a change event if the data actually changed
+		dataChangeEvent, err = persistDataChangeEvent(
+			ctx, db, record.TableName(), uuid, parentUUID, beforeModel, afterModel)
```

Comment:
So, this is what makes difficult using an upsert and/or run a bulk operation because as I commented previously, updating one object at a time is the lowest approach in pgx.

The way I see it is, we only compare the fields that are exposed in the API model. I don't see you have defaults in the table for those. So, record already have those values. If find a way to determine when an upsert inserts or updates, then we could replace everything inside `persistObject` by an upsert. 

I found this https://stackoverflow.com/questions/39058213/differentiate-inserted-and-updated-rows-in-upsert-using-system-columns. But of course, this needs further investigation.

This is a good start. We could improve it later

Reply:
I agree, this should be an area to revisit later.   As discussed the other day, having a proper CDC mechanism would solve these issues since we could blindly do an upsert and handle the change event separately to compare the fields.  Bringing it a full CDC mechanism is too much complexity for what we actually need.  Implemented them with a simpler DB trigger would be easier but would pose portability issues, as well as issues scaling the number of servers so we'll have to consider those issues.   For now I think this will serve us well enough.